tizen-api/src/nnstreamer-single.c

```diff
@@ -23,39 +23,373 @@
+          model_path);
+      break;
+    case ML_NNFW_TENSORFLOW_LITE:
+      if (!g_str_has_suffix (model_path, ".tflite")) {
+        dloge ("The given model file [%s] has invalid extension.", model_path);
+        ret = NNS_ERROR_INVALID_PARAMETER;
+        goto error;
+      }
+
+      pipeline_desc =
+          g_strdup_printf
+          ("appsrc name=srcx ! tensor_filter framework=tensorflow-lite model=%s ! appsink name=sinkx",
+          model_path);
+      break;
+    default:
+      /** @todo Add other fw later. */
+      dloge ("The given nnfw is not supported.");
+      ret = NNS_ERROR_NOT_SUPPORTED;
+      goto error;
+  }
 
   /* 2. Determine hw */
+  /** @todo Now the param hw is ignored. (Supposed CPU only) Support others later. */
+
+  /* 3. Construct a pipeline */
+  ret = nns_pipeline_construct (pipeline_desc, &pipe);
+  if (ret != NNS_ERROR_NONE) {
+    /* Failed to construct pipeline. */
+    goto error;
+  }
 
-  /* 3. Determine input dimension ==> caps_filter string */
+  pipe_h = (nns_pipeline *) pipe;
+  appsrc = gst_bin_get_by_name (GST_BIN (pipe_h->element), "srcx");
+  appsink = gst_bin_get_by_name (GST_BIN (pipe_h->element), "sinkx");
 
-  /* 4. Construct a pipeline */
-  _model = g_new (ml_simpleshot_model, 1);
-  ret = nns_pipeline_construct (pipedesc, &_model->pipe);
+  /* Supposed input type is single tensor if the number of tensors is 1. */
+  if (config.info.num_tensors == 1) {
+    GstTensorConfig c;
+
+    gst_tensor_info_copy (&c.info, &config.info.info[0]);
+    c.rate_n = 0;
+    c.rate_d = 1;
+
+    caps = gst_tensor_caps_from_config (&c);
+    gst_tensor_info_free (&c.info);
+  } else {
+    caps = gst_tensors_caps_from_config (&config);
+  }
+
+  gst_app_src_set_caps (GST_APP_SRC (appsrc), caps);
+  gst_caps_unref (caps);
+
+  ret = nns_pipeline_start (pipe);
+  if (ret != NNS_ERROR_NONE) {
+    /* Failed to construct pipeline. */
+    goto error;
+  }
 
   /* 5. Allocate */
-  *model = _model;
+  model_h = g_new0 (ml_simpleshot_model, 1);
+  *model = model_h;
+
+  model_h->pipe = pipe;
+  model_h->src = appsrc;
+  model_h->sink = appsink;
+
+  /* Set in/out tensors info */
+  gst_tensors_info_copy (&model_h->in_info, &config.info);
+
+  caps = gst_app_sink_get_caps (GST_APP_SINK (appsink));
+  gst_tensors_config_from_structure (&config, gst_caps_get_structure (caps, 0));
+  gst_caps_unref (caps);
+
+  gst_tensors_info_copy (&model_h->out_info, &config.info);
+
+error:
+  if (pipeline_desc)
+    g_free (pipeline_desc);
+
+  return ret;
+}
+
+/**
+ * @brief Close the opened model handle. (more info in nnstreamer-single.h)
+ */
+int
+ml_model_close (ml_simpleshot_model_h model)
+{
+  ml_simpleshot_model *model_h;
+
+  if (!model) {
+    dloge ("The given param, model is invalid.");
+    return NNS_ERROR_INVALID_PARAMETER;
+  }
+
+  model_h = (ml_simpleshot_model *) model;
+
+  if (model_h->src)
+    gst_object_unref (model_h->src);
```

Comment:
Wouldn't it be safer with ```model_h->src = NULL;```  after unref?

Reply:
add code to set null after unref the element.